### PR TITLE
[bug_5548] Added feature: mean time starting and ending time work.

### PR DESCRIPTION
**Pull request by MarcinWyt**
Proszę o review.
https://gist.github.com/MarcinWyt/42459ea6103a1a38b7bb
https://bugzilla.stxnext.pl/show_bug.cgi?id=5548

### CODE REVIEW

File: src/presence_analyzer/static/presence_start_end.html [hunk 1]
```diff
@@ -96,33 +101,43 @@
                     if(selected_user) {
                         loading.show();
                         chart_div.hide();
-
+
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: Delete white space please.

File: src/presence_analyzer/static/presence_start_end.html [hunk 2]
```diff
@@ -96,33 +101,43 @@
...
-                        chart.draw(data, options);
-
+                        // debugger
+                        // var result = [
+                        //     ["Mon", new Date(1, 1, 1, 7, 37, 38), new Date(1, 1, 1, 15, 07, 14)],
+                        //     ["Tue", new Date(1, 1, 1, 8, 16, 23), new Date(1, 1, 1, 14, 01, 47)],
+                        //     ["Wed", new Date(1, 1, 1, 7, 00, 06), new Date(1, 1, 1, 13, 24, 12)],
+                        //     ["Thu", new Date(1, 1, 1, 7, 26, 32), new Date(1, 1, 1, 12, 40, 45)],
+                        //     ["Fri", new Date(1, 1, 1, 6, 53, 22), new Date(1, 1, 1, 12, 52, 09)]
+                        // ]
```
Comment (dpolcyn) [inline src/presence_analyzer/static/presence_start_end.html]: We don't need this.

File: src/presence_analyzer/tests.py [hunk 3]
```diff
@@ -223,6 +240,32 @@ def test_mean(self):
         result = utils.mean(items)
         self.assertEqual(result, 2.975)
 
+    def test_list_start_end_time(self):
+        """
+        Test creating list starting and ending work time for specyfic user
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Full stop at the end of the line.

File: src/presence_analyzer/tests.py [hunk 4]
```diff
@@ -223,6 +240,32 @@ def test_mean(self):
...
+        expected_data = {
+            0: {'start': [], 'end': []},
+            1: {'start': [34745], 'end': [64792]},
+            2: {'start': [33592], 'end': [58057]},
+            3: {'start': [38926], 'end': [62631]},
+            4: {'start': [], 'end': []},
+            5: {'start': [], 'end': []},
+            6: {'start': [], 'end': []}
+        }
+        a_data = {
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: Just use ```data```.

File: src/presence_analyzer/static/presence_start_end.html [hunk 5]
```diff
@@ -96,33 +101,43 @@
                     if(selected_user) {
                         loading.show();
                         chart_div.hide();
-
+                        
                         // some dummy data, this should go from API
```
Comment (mjelonek92) [inline src/presence_analyzer/static/presence_start_end.html]: Delete this comment please.

File: src/presence_analyzer/utils.py [hunk 6]
```diff
@@ -102,3 +102,21 @@ def mean(items):
     Calculates arithmetic mean. Returns zero for empty lists.
     """
     return float(sum(items)) / len(items) if len(items) > 0 else 0
+
+
+def list_start_end_time(items):
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: List is not descriptive enough. This function groups ```items``` by weekday, and this is what the function name should say, e.g.: ```group_start_end_times_by_weekday```

File: src/presence_analyzer/views.py [hunk 7]
```diff
@@ -4,10 +4,13 @@
 """
 
 import calendar
+import datetime
 from flask import redirect, abort
 
 from presence_analyzer.main import app
-from presence_analyzer.utils import jsonify, get_data, mean, group_by_weekday
+from presence_analyzer.utils import \
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: Style. Please use:
```
from presence_analyzer.utils import (
    jsonify,
    get_data,
    mean,
    group_by_weekday,
    list_start_end_time,
)
```

File: src/presence_analyzer/views.py [hunk 8]
```diff
@@ -73,3 +76,29 @@ def presence_weekday_view(user_id):
...
+def presence_start_end_view(user_id):
+    """
+    Returns mean start time and mean end time.
+    """
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    aweek = {}
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: You don't need to declare it.

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -94,6 +94,23 @@ def test_presence_weekday_view(self):
...
         self.assertEqual(resp_data, expected_data)
 
+    def presence_start_end_view(self):
+        """
+        Test mean starting and ending time.
+        """
+        resp = self.client.get('/api/v1/presence_start_end/10')
+        self.assertEqual(resp.status_code, 200)
+        resp_data = json.loads(resp.data)
+        expected_data = [['Mon', 0, 0],
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: We missed this before, but please correct this style to match:
```
expected_data = [
    ['Mon', 0, 0],
    ['Tue', 34745.0, 64792.0],
    ...
]
```

If you could fix this in other places too, thanks!

File: src/presence_analyzer/views.py [hunk 10]
```diff
@@ -73,3 +81,28 @@ def presence_weekday_view(user_id):
...
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    aweek = group_start_end_times_by_weekday(data[user_id])
+
+    result = []
+    for aday in aweek:
+        starts = mean(aweek[aday]['start'])
+        ends = mean(aweek[aday]['end'])
+        result.append([aday, starts, ends])
```
Comment (mjelonek92) [inline src/presence_analyzer/views.py]: Can't you append with calendar name already? Like so: ```calendar.day_abbr[aday]```?

File: src/presence_analyzer/utils.py [hunk 11]
```diff
@@ -102,3 +102,21 @@ def mean(items):
...
+    """
+    user_week = {
+        i: {'start': [], 'end': []}
+        for i in range(7)
+    }
+    for item in items:
+        start = items[item]['start']
+        end = items[item]['end']
+        user_week[item.weekday()]['start'].\
+            append(seconds_since_midnight(start))
```
Comment (mjelonek92) [inline src/presence_analyzer/utils.py]: Perhaps?
```python
user_week[item.weekday()]['start'].append(
    seconds_since_midnight(start)
)
```

**Comment by dpolcyn** [comment]
Squash please.

File: src/presence_analyzer/views.py [hunk 12]
```diff
@@ -73,3 +81,25 @@ def presence_weekday_view(user_id):
...
+    """
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    week = group_start_end_times_by_weekday(data[user_id])
+
+    result = []
+    for aday in week:
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: day, not aday

File: src/presence_analyzer/tests.py [hunk 13]
```diff
@@ -84,14 +91,36 @@ def test_presence_weekday_view(self):
...
+            ["Wed", 24465],
+            ["Thu", 23705],
+            ["Fri", 0],
+            ["Sat", 0],
+            ["Sun", 0]
+        ]
+
+        self.assertEqual(resp_data, expected_data)
+
+    def presence_start_end_view(self):
```
Comment (mjelonek92) [inline src/presence_analyzer/tests.py]: Test method must start with ```test_```.

**Comment by dpolcyn** [comment]
Squash please.

File: src/presence_analyzer/static/presence_start_end.html [hunk 14]
```diff
@@ -95,34 +100,31 @@
...
-                        chart.draw(data, options);
-
+                        chart_div.hide();                        
+                        $.getJSON("/api/v1/presence_start_end/"+selected_user, 
+                            function(result) {
+                                $.each(result, function( index, value ) {
+                                    value[1] = parseInterval(value[1]);
+                                    value[2] = parseInterval(value[2]);
+                                })
+                                console.log(result);
```
Comment (mjelonek92) [inline src/presence_analyzer/static/presence_start_end.html]: Remove this please.

**Comment by dpolcyn** [comment]
Squash please.

**Comment by dpolcyn** [comment]
Change commit name. ```Some bugs removed.``` doesn't give any information about new view.

**Pull request merged**